txp/Pos/Txp/Logic/Global.hs

```diff
@@ -25,116 +26,193 @@ import           Pos.Core.Block.Union (ComponentBlock (..))
-    }
+----------------------------------------------------------------------------
+-- General processing
+----------------------------------------------------------------------------
 
-applyBlocksSettings
-    :: GlobalApplyToilMode m
-    => ApplyBlocksSettings () m
-applyBlocksSettings =
-    ApplyBlocksSettings
-    { absApplySingle = applyToil . blundToAuxNUndo
-    , absExtraOperations = const mempty
+data ProcessBlundsSettings extraEnv extraState m = ProcessBlundsSettings
+    { pbsProcessSingle   :: TxpBlund -> m (ExtendedGlobalToilM extraEnv extraState ())
+    , pbsCreateEnv       :: Utxo -> [TxAux] -> m extraEnv
+    , pbsExtraOperations :: extraState -> SomeBatchOp
+    , pbsIsRollback      :: !Bool
+    -- ^ This flag specifies whether we want to rollback transactions
+    -- or apply them. It affects the way we construct base 'Utxo'. If
+    -- we want to apply transactions, we should use 'buildUtxo' to
+    -- resolved all their inputs. But if we want to rollback them, we
+    -- should turn known outputs of transactions into 'Utxo'.
     }
 
-applyBlocksWith
-    :: (TxpGlobalApplyMode ctx m, Default extra)
-    => ApplyBlocksSettings extra (ToilT extra (DBToil m))
+processBlunds ::
+       forall extraEnv extraState m. (TxpCommonMode m, Default extraState)
+    => ProcessBlundsSettings extraEnv extraState m
+    -> NE TxpBlund
+    -> m SomeBatchOp
+processBlunds ProcessBlundsSettings {..} blunds = do
+    let toBatchOp (gts, extra) =
+            globalToilStateToBatch gts <> pbsExtraOperations extra
+    totalStake <- DB.getRealTotalStake -- doesn't change
+    -- Note: base utxo also doesn't change, but we build it on each
+    -- step (for different sets of transactions), because
+    -- 'UtxoModifier' may accumulate some data and it may be more
+    -- efficient.
+
+    -- Another note: if we rollback transactions, we don't really need
+    -- base utxo, but we have a sanity check in 'utxoDel' which forces
+    -- us to construct base utxo here.
+    let buildBaseUtxo :: UtxoModifier -> [TxAux] -> m Utxo
+        buildBaseUtxo
+            | pbsIsRollback = buildUtxoForRollback
+            | otherwise = buildUtxo
+
+    let step ::
+               (GlobalToilState, extraState)
+            -> TxpBlund
+            -> m (GlobalToilState, extraState)
+        step st txpBlund = do
+            processSingle <- pbsProcessSingle txpBlund
+            let txAuxesAndUndos = blundToAuxNUndo txpBlund
+                txAuxes = fst <$> txAuxesAndUndos
+            baseUtxo <- buildBaseUtxo (st ^. _1 . gtsUtxoModifier) txAuxes
+            extraEnv <- pbsCreateEnv baseUtxo txAuxes
+            let gte =
+                    GlobalToilEnv
+                        { _gteUtxo = utxoToLookup baseUtxo
+                        , _gteTotalStake = totalStake
+                        }
+            let env = (gte, extraEnv)
+            runGlobalToilMBase DB.getRealStake . flip execStateT st .
```

Comment:
`Universum.executingStateT` perhaps?

Reply:
I am not a hater of `flip` and `flip execStateT` is not even longer and it's more familiar to everybody, so I prefer `flip execStateT`.